[Auditbeat] Add user metricset

cwurm opened the pull request:
Adds a list of users with all their information to the `host` metricset. Works on Linux and macOS by using `getpwent(3)`.

ruflin commented on x-pack/auditbeat/module/system/host/_meta/data.json:
```diff
@@ -40,7 +40,18 @@
             "timezone.offset.sec": 0,
-            "uptime": 222611977726
+            "uptime": 222611977726,
+            "users": [
```
I'm worried about adding nested objects to our data structure. What we normally do in these cases send separate events for each entry.

andrewkroh replied:
This would be another place where not using a nested document would allow us to map some of the fields to [ECS](https://github.com/elastic/ecs#user) (like user.id and user.name).

ruflin commented on x-pack/auditbeat/module/system/host/_meta/fields.yml:
```diff
@@ -95,3 +95,36 @@
         description: >
           IP addresses.
+  - name: users
+    type: array
```
I don't think this will have your expected result in the template

andrewkroh commented on x-pack/auditbeat/module/system/host/users_unix.go:
```diff
@@ -0,0 +1,65 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build !windows
```
This should also a have a `cgo` build tag (like `!windows,cgo`). I haven't tried this in a while with auditbeat but it's nice if you can still build the project without cgo (CGO_ENABLED=false) even it it has limited features.

andrewkroh commented on x-pack/auditbeat/module/system/host/host.go:
```diff
@@ -54,6 +54,19 @@ func (ms *MetricSet) Fetch(report mb.ReporterV2) {
 		return
 	}
 
+	users, err := GetUsers()
```
You'll need a `GetUsers` stub with a build tag like `// +build windows !cgo` (meaning `windows OR (not CGO))`).

andrewkroh commented on x-pack/auditbeat/module/system/host/_meta/fields.yml:
```diff
@@ -95,3 +95,36 @@
+      type: keyword
+      description: >
+          Information about a user's password.
+    - name: dir
```
I think I would start with `keyword` for `dir` and `shell` then add an analyzed version later if requested.

andrewkroh commented on x-pack/auditbeat/module/system/host/users_unix.go:
```diff
@@ -0,0 +1,65 @@
+		"passwd": user.passwd,
+		"uid":    user.uid,
+		"gid":    user.gid,
+		"gecos":  user.gecos,
```
Can some of these values be empty strings? I was thinking that some of them are not always populated so you should drop those keys.

andrewkroh commented on x-pack/auditbeat/module/system/host/_meta/fields.yml:
```diff
@@ -95,3 +95,36 @@
+      type: text
+      description: >
+          Program to run at login.
+    - name: gecos
```
Could we call this something generic like `user_information`? I find `gecos` to be very unintuitive. Looks like the term refers to a really old OS https://en.wikipedia.org/wiki/General_Comprehensive_Operating_System.

cwurm commented:
Thanks, @andrewkroh and @ruflin. I've (hopefully) addressed all comments for now. This is now its own metricset, and sending each user as its own document.

I'm still working on adding persistence across Auditbeat restarts, and we have some outstanding naming discussions (I put `event.type` / `event.type_detail` for now, but those are TBD).

So I'm keeping the PR in `in progress` for now, and you don't need to do another review at this point.

andrewkroh commented on x-pack/auditbeat/module/system/user/user.go:
```diff
@@ -0,0 +1,303 @@
+		}
+	}
+
+	ms.bucket.Store("users", buf.Bytes())
```
This returns an error that is unhandled. Try using [gometalinter](https://github.com/alecthomas/gometalinter). It will run `errcheck` for you which points these kind of things out.

cwurm replied:
Thanks - `gometalinter` is a handy tool. It's a shame that `errcheck` finds so many problems in the Beats codebase. Most commonly, we don't check the error returns of calls to `common.MapStr.Put` and `XXHash64.Write*`. Do you think we should be checking them, or `/nolint`-ing?

andrewkroh replied:
IIRC XXHash64.Write does not ever return an error, but it has that signature to satisfy the `hash.Hash` interface. Likewise there are cases where MapStr.Put can never return an error like `Put("x", 1)` so skipping the check is fine.

If we were going to start linting all the time then yes figuring a way to squelch the false positives would be good. There's a tool called reviewdog (which `make lint`) uses that is supposed to only alert you to the linter warnings that were introduced on your branch. It compares the results from `master` against the current branch.

andrewkroh commented on x-pack/auditbeat/module/system/user/user.go:
```diff
@@ -0,0 +1,303 @@
+		}
+	}
+
+	ms.bucket.Store("users", buf.Bytes())
```
After reporting that a user was removed I think you should delete the old entry from the bucket to prevent it from growing without bound.

cwurm replied:
The list of users is written to disk - overwriting the previous list - at the end of `Fetch`. This happens only if there was a change, and at least for users that should not be that common.

andrewkroh replied:
Oh, I see the key is `users`. 👍

andrewkroh commented on x-pack/auditbeat/module/system/user/user.go:
```diff
@@ -0,0 +1,303 @@
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet(moduleName, metricsetName, New,
```
If someone tries using this on Windows ideally they would get an error at startup or on `test config`. In order to make that happen you need to do one of these

- register a different factory function for Windows just returns a friendly error (use build tags to do this conditionally)
- not register the metricset on windows (gives a vague error)
- check for windows in this metricset factory function (`New`) and return an error

cwurm replied:
Makes sense - I've added a check in `New`.

cwurm commented on x-pack/auditbeat/module/system/user/users_unix.go:
```diff
@@ -0,0 +1,40 @@
+		// passwd is C.struct_passwd
+		users = append(users, &User{
+			name:     C.GoString(passwd.pw_name),
+			passwd:   C.GoString(passwd.pw_passwd),
```
In older days this could contain the encrypted password (which we might indeed not want to send), but all modern UNIX systems set this to either `x` (password is in the shadow file) or `*` (login disabled). So I guess it depends on whether there is value in knowing whether the login for a user is disabled or not.

andrewkroh replied:
I think it would be nice to translate them to some friendly values like `login_disabled` (*) or `in_shadow_file` (x) or `redacted` (all other values). Those would be more meaningful to me. WDYT?

cwurm replied:
Makes sense, I've changed the possible values to `shadow_password`, `login_disabled`, `no_password` (which is a dangerous vulnerability since anybody can log in), and `<redacted>`.

cwurm commented on x-pack/auditbeat/module/system/user/_meta/data.json:
```diff
@@ -0,0 +1,24 @@
+    "event.action": "existing_user",
+    "event.type": "state",
+    "system": {
+        "user": {
```
But keep them under `system/user` as well? I.e.:

```
"user": {
  "uid": 1001,
  "gid": 1001,
  "name": "ubuntu"
},
"system": {
    "user": {
      "passwd": "x",
      "uid": 1001,
      "gid": 1001,
      "dir": "/home/ubuntu",
      "shell": "/bin/bash",
      "user_information": "Ubuntu",
      "name": "ubuntu"
    }
  },
```

andrewkroh replied:
@ruflin @webmat Would you rather have some of the fields duplicated to `user.*` or should we move/promote the ECS fields (no duplication)?

cwurm replied:
Ok, should the final result look like this:

```
"user": {
  "id": 1001,
  "name": "ubuntu"
},
"system": {
    "user": {
      "passwd": "x",
      "gid": 1001,
      "dir": "/home/ubuntu",
      "shell": "/bin/bash",
      "user_information": "Ubuntu",
    }
  },
```

Only `user.id` and `user.name` currently seem to be in ECS. `uid`/`gid` are not. Personally, I would prefer to use `uid` since it's such a common term.

webmat replied:
Copying data over or aliasing are both fine by me, I don't have a strong opinion on this part.

I absolutely think it's important to fill the ECS fields as much as possible, however :-)

Regarding the field name, `user.id` would indeed be the place to put the Linux uid at this time. I understand that in the context of Linux, uid and gid is the go to name for these values. However I don't see that as reason enough to define a second field under `user` (and eventually `group`). Other arbitrary systems (OS or otherwise) will rarely use this nomenclature, so this would give us a situation where almost all systems fill `user.id` and exactly one system -- Linux -- fills `user.uid`.

I'm totally ok if you use uid/gid under the `system_audit` namespace, however.

cwurm replied:
To me at least the meaning of `user.id` would not obvious. If I'm a *NIX admin, I would wonder whether this is a custom ID generated by  Elastic Stack (for example, to identify the same user across systems where on each system they might have a different `uid`) rather than my well-known `uid`. As a Windows admin, I would wonder whether this is my `SID` or not. In any case, it's not obvious because it doesn't follow convention.

So I'd want to keep `system.user.uid` (and probably have a `system.user.sid` in the future for Windows). We can copy it to a top-level `user.id` (if we don't want to use that for some sort of cross-system identification).

cwurm replied:
Sounds good. We now have a top-level `user.id` and a `system.user.uid` field that share the same value: https://github.com/elastic/beats/blob/29db6420babaa534416c968350103acaeb6ac01b/x-pack/auditbeat/module/system/user/_meta/data.json

cwurm commented on x-pack/auditbeat/module/system/user/user.go:
```diff
@@ -0,0 +1,303 @@
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+
+	"github.com/OneOfOne/xxhash"
```
Ok. Shouldn't `make fmt` do this kind of thing?

andrewkroh replied:
`goimports` tries to be flexible. If you were to fully remove all the newlines from the import block it would do the right thing.

cwurm replied:
👍

cwurm commented on x-pack/auditbeat/module/system/user/user.go:
```diff
@@ -0,0 +1,303 @@
+	return ms, nil
+}
+
+// Load user cache from disk.
```
Makes sense.

cwurm commented on x-pack/auditbeat/module/system/user/user.go:
```diff
@@ -0,0 +1,303 @@
+	err = ms.bucket.Load("users", func(blob []byte) error {
+		if len(blob) > 0 {
+			buf := bytes.NewBuffer(blob)
+			decoder = gob.NewDecoder(buf)
```
I think that would work. It would force the full list of users to be sent to ES again, but that's ok.

cwurm commented:
I think this is ready for another review.

The `user` metricset now does everything that it should do. It reads (via C functions) `/etc/passwd`, `/etc/shadow`, and `/etc/group` to collect all relevant information about a user.

It can detect new users, deleted users, changes to users (e.g. groups), and - as a special distinct category - password changes.

Periodically, it sends all user information as state (frequency controlled by the config values `state.period` and `user.state.period` - default `12h`). Otherwise, it periodically (controlled by the usual `period` config - very low values are possible and recommended since when no changes have occurred it only triggers three syscalls) checks the `ctime` of the above files, reads them if the `ctime` has changed, detects changes compared to its internal cache, and reports changes.

The cache is persisted to disk in a `beat.db` file (already used by the `file_integrity` module) after every `Fetch` and on `Close`. It contains all current user information incl. a 10-character excerpt of the password hash from `/etc/shadow` (necessary to detect any password changes incl. a direct edit of `/etc/shadow` - however, that excerpt is not output to ES or any other output).

Possible future enhancements:

- [ ] Use `inotify` instead of periodic `stat` calls to detect possible changes.
- [ ] Read `/etc/sudoers` to add sudo information to users.

webmat commented on x-pack/auditbeat/module/system/user/user.go:
```diff
@@ -0,0 +1,431 @@
+	if err != nil {
+		return errors.Wrap(err, "failed to get users")
+	}
+	ms.log.Debugf("Found %v users", len(users))
```
Total user count could also be a good shorthand to save on a checkin / snapshot.

cwurm replied:
You mean send it in every user document? I guess we could, but maybe also here we could wait until we have an clear need (i.e. the UI would benefit from it).

webmat replied:
Hmmm, I may have assumed something from our original discussions. In my mind, there's one document being sent with overall details about a "checkin", and multiple associated documents, one per user, one per process, etc. So I would have saved that user count (and group count for that matter) on that one checkin document.

But thinking back about my review, I don't remember seeing such a generic checkin document, you're right. So I guess disregard for now. But it may be useful metadata, if we ever decide we need a single "checkin" document, to tie all other documents around it in the future.

webmat commented on x-pack/auditbeat/module/system/user/user_test.go:
```diff
@@ -0,0 +1,34 @@
+	mbtest "github.com/elastic/beats/metricbeat/mb/testing"
+)
+
+func TestData(t *testing.T) {
```
Should a function to generate test data be shipped as part of Auditbeat?

cwurm replied:
I'm not sure what you mean - this function does generate test data. Do you mean we shouldn't (most other metricsets have this) or that it should be different?

webmat replied:
Oh jeez, I missed the _test part of the file name 🤦‍♂️  Never mind, I'm not used to the Go conventions of having test files right besides their code file.

tsg commented on x-pack/auditbeat/module/system/user/users_linux.go:
```diff
@@ -68,6 +85,51 @@ func GetUsers() (users []*User, err error) {
+		*/
+		for i := 0; ; i++ {
+			offset := (unsafe.Sizeof(unsafe.Pointer(*cgroup.gr_mem)) * uintptr(i))
+			member := *(**C.char)(unsafe.Pointer(uintptr(unsafe.Pointer(cgroup.gr_mem)) + offset))
```
Tough one, but LGTM from what I can tell. 👍

cwurm replied:
Yeah, it's not pretty.

cwurm commented on x-pack/auditbeat/module/system/user/_meta/fields.yml:
```diff
@@ -0,0 +1,56 @@
+      type: text
+      description: >
+          General user information. On Linux, this is the gecos field.
+    - name: group
```
Unless we see an immediate need to store a count I'd want to pass. It's always possible to use a simple Lucene expression (something like `doc['system']['user']['group'].length`) in a query to get that value computed on the fly (and Lucene expressions are fast).

webmat replied:
We can do this via an ES API call? If that's so, then we can skip this counter field, I agree. And also I'd love to know how to do that :-)

cwurm replied:
Easiest way is to use a scripted field and access `_source`:

```
GET auditbeat-test/_search
{
    "query" : {
        "match_all": {}
    },
    "script_fields" : {
        "group_count" : {
            "script" : "params['_source']['system']['user']['group'].length"
        }
    }
}
```

andrewkroh commented on x-pack/auditbeat/module/system/_meta/config.yml.tmpl:
```diff
@@ -1,14 +1,17 @@
     - host
     - packages
     - processes
+    - user
```
The inclusion of `user` in the config file should be conditional on the `.GOOS` value. We don't want it showing up in packages on non-Linux.

cwurm replied:
Done.

andrewkroh commented on x-pack/auditbeat/module/system/user/_meta/fields.yml:
```diff
@@ -0,0 +1,56 @@
+    - name: password
+      type: group
+      description: >
+        `password` contains information about a user's password (not the password itself).
```
Can you also document what the possible values are for this enum.

cwurm replied:
Done.

cwurm commented:
After a conversation with @andrewkroh I changed the password change detection to use a SHA-512 hash of the full encrypted password (rather than the last 10 characters) - but excluding the salt. This should make sure that we detect any possible password change.

I merged `master` into the feature branch and rebased this PR, so the commits are now displayed all at the end. There are two commits since the last review: [`Switch to new UUID library.`](https://github.com/elastic/beats/pull/8835/commits/526880692d544432494c4122c6d2a8b03ba53047) and [
Use full encrypted password for change detection, store as SHA-512 hash. ](https://github.com/elastic/beats/pull/8835/commits/068130050460b99f325dc052d2516e574e8ccd96).

@andrewkroh let me know if this still looks good so I can merge.

andrewkroh commented on x-pack/auditbeat/module/system/user/users_linux.go:
```diff
@@ -150,24 +167,33 @@ func readShadowFile() (map[string]shadowFileEntry, error) {
+// and returns a SHA-512 hash of the rest.
+func noSaltHash(password string) []byte {
+	// On modern UNIX systems, the format of the password field is $id$salt$encrypted (see crypt(3)).
+	// We only want to use the `encrypted` part.
```
Why do we only want the encrypted part?

cwurm replied:
I was aiming to reduce the amount of sensitive data we hold on to (even in hashed form). The encrypted part is the hash that should be unique on its own and so should be good enough for our purpose of detecting a change.

andrewkroh replied:
Given that you are hashing the the entire value I don't see any reason to filter out the salt. Plus, IIUC, by filtering it out your lose the ability to detect changes to the salt. The seems like small amount of logic that could be dropped.

cwurm replied:
> Plus, IIUC, by filtering it out your lose the ability to detect changes to the salt.

I don't think the salt can change without the hash changing, too - given it's a part of the input to the hashing function?

I agree that filtering out the salt provides debatable benefits: An attacker would have to gain access to `beat.db`, successfully reverse the SHA-512 hash, and then compute a rainbow table for that specific salt value. But if we don't need the salt to detect a change - why take even a theoretical risk?

andrewkroh replied:
If the salt is changed in the file without changing the password hash the user will not be able to log in. At login the machine will compare the stored password hash with the output of `hash(salt+user_provided_password)`. Changing only the salt is not something that should happen, but I'd like to know if any part of the value changed (salt, algorithm, or hash).

> successfully reverse the SHA-512 hash

Unlike encryption, hashing is not reversible so this isn't a possibility.

cwurm replied:
> > successfully reverse the SHA-512 hash
> 
> Unlike encryption, hashing is not reversible so this isn't a possibility.

Is it never? I assume most collision attacks find inputs that produce the same hash, but are not actually the same input as the original that produced the hash in the first place (i.e. `hash(m1) == hash(m2)` but `m1 != m2`). But is that always the case, or could `m1 == m2`?

cwurm replied:
For posterity: We talked about this and agreed that there is little risk in storing a SHA-512 hash of the password field in a local database file that's only accessible by `root` (same as `/etc/shadow`).
